File: src/cubyz/world/terrain/ClimateMap.java
package cubyz.world.terrain;

import cubyz.utils.datastructures.Cache;
import cubyz.world.World;

public class ClimateMap {
	private static final int CACHE_SIZE = 1 << 8; // Must be a power of 2!
	private static final int CACHE_MASK = CACHE_SIZE - 1;
	private static final int ASSOCIATIVITY = 4;
	private static final Cache<ClimateMapFragment> cache = new Cache<ClimateMapFragment>(new ClimateMapFragment[CACHE_SIZE][ASSOCIATIVITY]);
	private static World world = null;
	public static BiomePoint[][] getBiomeMap(World world, int wx, int wz, int width, int height) {
		if (world != ClimateMap.world) {
			cache.clear(); // Clear the cache if the world changed!
			ClimateMap.world = world;
		}
		
		BiomePoint[][] map = new BiomePoint[width/MapFragment.BIOME_SIZE][height/MapFragment.BIOME_SIZE];
		int wxStart = wx & ~ClimateMapFragment.MAP_MASK;
		int wzStart = wz & ~ClimateMapFragment.MAP_MASK;
		int wxEnd = wx+width & ~ClimateMapFragment.MAP_MASK;
		int wzEnd = wz+height & ~ClimateMapFragment.MAP_MASK;
		for(int x = wxStart; x <= wxEnd; x += ClimateMapFragment.MAP_SIZE) {
			for(int z = wzStart; z <= wzEnd; z += ClimateMapFragment.MAP_SIZE) {
				ClimateMapFragment mapPiece = getOrGenerateFragment(world, x, z);
				// Offset of the indices in the result map:
				int xOffset = (x - wx) >> MapFragment.BIOME_SHIFT;
				int zOffset = (z - wz) >> MapFragment.BIOME_SHIFT;
				// Go through all indices in the mapPiece:
				for(int lx = 0; lx < mapPiece.map.length; lx++) {
					int resultX = lx + xOffset;
					if (resultX < 0 || resultX >= map.length) continue;
					for(int lz = 0; lz < mapPiece.map[0].length; lz++) {
						int resultZ = lz + zOffset;
						if (resultZ < 0 || resultZ >= map.length) continue;
						map[resultX][resultZ] = mapPiece.map[lx][lz];
					}
				}
			}
		}
		return map;
	}
	
	private static class ClimateMapFragmentComparator {
		private final int wx, wz;
		private ClimateMapFragmentComparator(int wx, int wz) {
			this.wx = wx;
			this.wz = wz;
		}
		@Override
		public boolean equals(Object other) {
			if (other instanceof ClimateMapFragment) {
				return ((ClimateMapFragment)other).wx == wx && ((ClimateMapFragment)other).wz == wz;
			}
			return false;
		}
	}
	
	public static ClimateMapFragment getOrGenerateFragment(World world, int wx, int wz) {
		int hash = ClimateMapFragment.hashCode(wx, wz) & CACHE_MASK;
		ClimateMapFragment ret = cache.find(new ClimateMapFragmentComparator(wx, wz), hash);
		if (ret != null) return ret;
		synchronized(cache.cache[hash]) {
			// Try again in case it was already generated in another thread:
			ret = cache.find(new ClimateMapFragmentComparator(wx, wz), hash);
			if (ret != null) return ret;
			ret = new ClimateMapFragment(world, wx, wz);
			world.chunkManager.climateGenerator.generateMapFragment(ret);
			cache.addToCache(ret, ret.hashCode() & CACHE_MASK);
			return ret;
		}
	}
}

File: src/server/terrain/ClimateMap.zig
const std = @import("std");
const Atomic = std.atomic.Value;

const main = @import("main");
const Array2D = main.utils.Array2D;
const Cache = main.utils.Cache;
const ZonElement = main.ZonElement;
const terrain = main.server.terrain;
const TerrainGenerationProfile = terrain.TerrainGenerationProfile;
const Biome = terrain.biomes.Biome;
const MapFragment = terrain.SurfaceMap.MapFragment;
const NeverFailingAllocator = main.heap.NeverFailingAllocator;

pub const climate_generators = @import("climategen/_list.zig");

pub const BiomeSample = struct {
	biome: *const Biome,
	height: f32,
	roughness: f32,
	hills: f32,
	mountains: f32,
	seed: u64,
};

const ClimateMapFragmentPosition = struct {
	wx: i32,
	wy: i32,

	pub fn equals(self: ClimateMapFragmentPosition, other: anytype) bool {
		if (@TypeOf(other) == ?*ClimateMapFragment) {
			if (other) |ch| {
				return self.wx == ch.pos.wx and self.wy == ch.pos.wy;
			}
			return false;
		} else @compileError("Unsupported");
	}

	pub fn hashCode(self: ClimateMapFragmentPosition) u32 {
		return @bitCast((self.wx >> ClimateMapFragment.mapShift)*%33 +% (self.wy >> ClimateMapFragment.mapShift));
	}
};

pub const ClimateMapFragment = struct {
	pub const mapShift = 8 + MapFragment.biomeShift;
	pub const mapSize = 1 << mapShift;
	pub const mapMask: i32 = mapSize - 1;

	pub const mapEntrysSize = mapSize >> MapFragment.biomeShift;

	pos: ClimateMapFragmentPosition,
	map: [mapEntrysSize][mapEntrysSize]BiomeSample = undefined,

	pub fn init(self: *ClimateMapFragment, wx: i32, wy: i32) void {
		self.* = .{
			.pos = .{.wx = wx, .wy = wy},
		};
	}

	fn privateDeinit(self: *ClimateMapFragment) void {
		memoryPool.destroy(self);
	}

	pub fn deferredDeinit(self: *ClimateMapFragment) void {
		main.heap.GarbageCollection.deferredFree(.{.ptr = self, .freeFunction = main.meta.castFunctionSelfToAnyopaque(privateDeinit)});
	}

	pub fn hashCode(wx: i32, wy: i32) u32 {
		return @bitCast((wx >> mapShift)*%33 + (wy >> mapShift));
	}
};

/// Generates the climate(aka Biome) map, which is a rough representation of the world.
pub const ClimateMapGenerator = struct {
	init: *const fn (parameters: ZonElement) void,
	generateMapFragment: *const fn (fragment: *ClimateMapFragment, seed: u64) void,

	const generatorRegistry: std.StaticStringMap(ClimateMapGenerator) = .initComptime(blk: {
		const decls = @typeInfo(climate_generators).@"struct".decls;
		var generators: [decls.len]struct { []const u8, ClimateMapGenerator } = undefined;
		for (0..decls.len) |i| {
			const Generator = @field(climate_generators, decls[i].name);
			generators[i] = .{Generator.id, .{
				.init = &Generator.init,
				.generateMapFragment = &Generator.generateMapFragment,
			}};
		}
		break :blk generators;
	});

	pub fn getGeneratorById(id: []const u8) !ClimateMapGenerator {
		return generatorRegistry.get(id) orelse {
			std.log.err("Couldn't find climate map generator with id {s}", .{id});
			return error.UnknownClimateMapGenerator;
		};
	}
};

const cacheSize = 1 << 5; // Must be a power of 2!
const cacheMask = cacheSize - 1;
const associativity = 8; // ~400 MiB
var cache: Cache(ClimateMapFragment, cacheSize, associativity, ClimateMapFragment.deferredDeinit) = .{};
var profile: TerrainGenerationProfile = undefined;

var memoryPool: main.heap.MemoryPool(ClimateMapFragment) = undefined;

pub fn globalInit() void {
	memoryPool = .init(main.globalAllocator);
}

pub fn globalDeinit() void {
	memoryPool.deinit();
}

fn cacheInit(pos: ClimateMapFragmentPosition) *ClimateMapFragment {
	const mapFragment = memoryPool.create();
	mapFragment.init(pos.wx, pos.wy);
	profile.climateGenerator.generateMapFragment(mapFragment, profile.seed);
	return mapFragment;
}

pub fn init(_profile: TerrainGenerationProfile) void {
	profile = _profile;
}

pub fn deinit() void {
	cache.clear();
}

pub fn getOrGenerateFragment(wx: i32, wy: i32) *ClimateMapFragment {
	const compare = ClimateMapFragmentPosition{.wx = wx, .wy = wy};
	const result = cache.findOrCreate(compare, cacheInit, null);
	return result;
}

pub fn getBiomeMap(allocator: NeverFailingAllocator, wx: i32, wy: i32, width: u31, height: u31) Array2D(BiomeSample) {
	const map = Array2D(BiomeSample).init(allocator, width >> MapFragment.biomeShift, height >> MapFragment.biomeShift);
	const wxStart = wx & ~ClimateMapFragment.mapMask;
	const wzStart = wy & ~ClimateMapFragment.mapMask;
	const wxEnd = wx +% width & ~ClimateMapFragment.mapMask;
	const wzEnd = wy +% height & ~ClimateMapFragment.mapMask;
	var x = wxStart;
	while (wxEnd -% x >= 0) : (x +%= ClimateMapFragment.mapSize) {
		var y = wzStart;
		while (wzEnd -% y >= 0) : (y +%= ClimateMapFragment.mapSize) {
			const mapPiece = getOrGenerateFragment(x, y);
			// Offset of the indices in the result map:
			const xOffset = (x -% wx) >> MapFragment.biomeShift;
			const yOffset = (y -% wy) >> MapFragment.biomeShift;
			// Go through all indices in the mapPiece:
			for (&mapPiece.map, 0..) |*col, lx| {
				const resultX = @as(i32, @intCast(lx)) + xOffset;
				if (resultX < 0 or resultX >= width >> MapFragment.biomeShift) continue;
				for (col, 0..) |*spot, ly| {
					const resultY = @as(i32, @intCast(ly)) + yOffset;
					if (resultY < 0 or resultY >= height >> MapFragment.biomeShift) continue;
					map.set(@intCast(resultX), @intCast(resultY), spot.*);
				}
			}
		}
	}
	return map;
}
